osu.Game.Rulesets.Osu/Mods/OsuModTarget.cs

```diff
@@ -1,19 +1,528 @@
+
+                var avgColour = colour.AverageColour.Linear;
+                drawable.FadeColour(new Color4(avgColour.R * 0.45f, avgColour.G * 0.45f, avgColour.B * 0.45f, avgColour.A))
+                        .Then().Delay(h.TimePreempt - controlPointInfo.TimingPointAt(h.StartTime).BeatLength - undim_duration)
+                        .FadeColour(colour, undim_duration);
+
+                // remove approach circles
+                circle.ApproachCircle.Hide();
+            }
+        }
+
+        #endregion
+
+        #region Beatmap Generation (IApplicableToBeatmap)
+
+        public override void ApplyToBeatmap(IBeatmap beatmap)
+        {
+            Seed.Value ??= RNG.Next();
+
+            var osuBeatmap = (OsuBeatmap)beatmap;
+
+            if (osuBeatmap.HitObjects.Count == 0) return;
+
+            controlPointInfo = osuBeatmap.ControlPointInfo;
+            var origHitObjects = osuBeatmap.HitObjects.OrderBy(x => x.StartTime).ToList();
+
+            var hitObjects = generateBeats(osuBeatmap, origHitObjects)
+                             .Select(x =>
+                             {
+                                 var newCircle = new HitCircle();
+                                 newCircle.ApplyDefaults(controlPointInfo, osuBeatmap.BeatmapInfo.BaseDifficulty);
+                                 newCircle.StartTime = x;
+                                 return (OsuHitObject)newCircle;
+                             }).ToList();
+
+            addHitSamples(hitObjects, origHitObjects);
+
+            fixComboInfo(hitObjects, origHitObjects);
+
+            randomizeCirclePos(hitObjects);
+
+            osuBeatmap.HitObjects = hitObjects;
+
+            base.ApplyToBeatmap(beatmap);
+        }
+
+        private IEnumerable<double> generateBeats(IBeatmap beatmap, IReadOnlyCollection<OsuHitObject> origHitObjects)
+        {
+            var startTime = origHitObjects.First().StartTime;
+            var endObj = origHitObjects.Last();
+            var endTime = endObj switch
+            {
+                Slider slider => slider.EndTime,
+                Spinner spinner => spinner.EndTime,
+                _ => endObj.StartTime
+            };
+
+            var beats = beatmap.ControlPointInfo.TimingPoints
+                               .Where(x => Precision.AlmostBigger(endTime, x.Time))
+                               .SelectMany(tp =>
+                               {
+                                   var tpBeats = new List<double>();
+                                   var currentTime = tp.Time;
+
+                                   while (Precision.AlmostBigger(endTime, currentTime) && beatmap.ControlPointInfo.TimingPointAt(currentTime) == tp)
+                                   {
+                                       tpBeats.Add(Math.Floor(currentTime));
+                                       currentTime += tp.BeatLength;
+                                   }
+
+                                   return tpBeats;
```

Comment:
* This should really be a method rather than an inline linq lambda. And one with preferably longer variable names.
* This method of generating beats is susceptible to drifting out of sync due to accumulating floating point error. To avoid that this should be done via multiplication, e.g. `startTime + i * beatLength` where `i` increments up until the entire expression exceeds `endTime`.